klippy/extras/dockable_probe.py

```diff
@@ -0,0 +1,596 @@
+        self.toolhead.manual_move([approach[0], approach[1], None],
+                                  self.attach_speed)
+
+    # Determine toolhead distance to dock coordinates
+    def _check_distance(self, pos=None, dist=None):
+        if not pos:
+            pos = self.toolhead.get_position()
+        dock = self.dock_position
+
+        if dist > sqrt((pos[0]-dock[0])**2 +
+                       (pos[1]-dock[1])**2 ):
+            return True
+        else:
+            return False
+
+    # Find a point on a vector line at a specific distance
+    def _get_point_on_vector(self, point, angle, magnitude=1):
+        x = point[0] - magnitude * cos(angle)
+        y = point[1] - magnitude * sin(angle)
+        return (x, y)
+
+    # Locate the intersection of two vectors
+    def _get_intercept(self, point1, angle1, point2, angle2):
+        x1, y1 = point1[:2]
+        x2, y2 = self._get_point_on_vector(point1, angle1, 10.0)
+        x3, y3 = point2[:2]
+        x4, y4 = self._get_point_on_vector(point2, angle2, 10.0)
+        det1 = ((x1 * y2) - (y1 * x2))
+        det2 = ((x3 * y4) - (y3 * x4))
+        d = ((x1 - x2) * (y3 - y4)) - ((y1 - y2) * (x3 - x4))
+        x = float((det1 * (x3 - x4)) - ((x1 - x2) * det2)) / d
+        y = float((det1 * (y3 - y4)) - ((y1 - y2) * det2)) / d
+        return (x, y)
+
+    # Determine the vector of two points
+    def _get_vector(self, point1, point2):
+        x1, y1 = point1[:2]
+        x2, y2 = point2[:2]
+        magnitude = sqrt((x2-x1)**2 + (y2-y1)**2 )
+        angle = atan2(y2-y1, x2-x1) + pi
+
+        return angle, magnitude
+
+    # Align z axis to prevent crashes
+    def _align_z(self):
+        if not self._last_homed:
+            curtime = self.printer.get_reactor().monotonic()
+            homed_axes = self.toolhead.get_status(curtime)['homed_axes']
+            self._last_homed = homed_axes
+        if self.dock_fixed_z:
+            dock_z = (self.dock_position[2] +
+                      self.position_endstop +
+                      self.sample_retract_dist)
+            tpos = self.toolhead.get_position()
+            if 'z' in self._last_homed:
+                if tpos[2] < dock_z:
+                    self.toolhead.set_position(tpos)
+                    self.toolhead.manual_move(
+                        [None, None, dock_z],
+                        self.lift_speed)
+            else:
+                self._force_z_hop()
+        else:
+            if 'z' in self._last_homed:
+                self.toolhead.manual_move(
+                                        [None, None, self.dock_position[2]],
+                                        self.lift_speed)
+            else:
+                raise self.printer.command_error(
+                    "Cannot attach/detach probe, must home Z axis first")
+
+    # Hop z and return to un-homed state
+    def _force_z_hop(self):
+        this_z = self.toolhead.get_position()[2]
+        if self.last_z != this_z:
```

Comment:
Could this be rewritten as early return?

Reply:
yes